File: 143. Reorder List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reorderList(self, head: ListNode) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        if not head or not head.next:
            return
        # find the mid point
        slow = fast = head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
        # reverse the second part
        curr = slow
        prev = None
        while curr:
            temp = curr.next
            curr.next = prev
            prev = curr
            curr = temp
        # relink nodes
        node = head
        while prev.next and node.next:
            temp = node.next
            node.next = prev
            revtemp = prev.next
            prev.next = temp
            node = temp
            prev = revtemp

File: 143. Reorder List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public void reorderList(ListNode head) {
        if (head == null || head.next == null) {
            return;
        }
        ListNode fast=head, slow=head;
        while (fast != null && fast.next != null) {
            fast = fast.next.next;
            slow = slow.next;
        }
        ListNode prev=null, curr=slow, next;
        while (curr != null) {
            next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        ListNode n1=head, n2=prev;
        while (n1 != null) {
            next = n1.next;
            n1.next = n2;
            n1 = next;
            if (n2 == null) {
                break;
            }
            next = n2.next;
            n2.next = n1;
            n2 = next;
        }
    }
}